doc/design/012-libindy-architecture-overview/README.md

```diff
@@ -0,0 +1,71 @@
+
+### Command Layer
+
+In the command layer, this is where functions combine the the majority of their logic. The command layer consumes the service layers and passes data between different services as is needed rather than allowing services to communicate directly with each other. When a particular command is called by the API layer by the CommandExecutor which then calls internal commands for each individual function. A command module will call one or many services to execute a function and then return the results back up to the API layer and then passed back up to the wrapper layer. Along with routing the commands to their respective module calls, the CommandExecutor also handles the creation of the services that are consumed by the command modules.
+
+
+### Service Layer
+
+Within the service layer we have very low level functions that can be consumed by the command layer. These low level functions are common functions which may be needed in many different command modules. For example within the crypto service, there is a common function to create a cryptographic key with a common interface to allow for key creation to happen for many different cryptographic material. The current implementation supports the creation of ed25519 keys, but could easily add support for secp256k1 keys as well.
+
+Another example of the functions supported at the wallet service layer is the ability to take a key that was created and to be able to add it to a database that's exposed through a common wallet interface. This common wallet interface allows for different DBs to be plugged in to the wallet interface depending on the needs of the application.
+
+In summary, the service layer can be thought of as a suite of helper primitives that will be consumed many times by the command layer. This helps libindy to adhere to the DRY (Don't repeat yourself) programming paradigm. Another advantage of the service layer is that it creates a clean, modular approach to quickly add new functionality at the API and command layers.
+
+### Errors
+
+This is where errors are specified in rust types which get translated at the API layer into errorcodes. Since only simple native c-types are used, libindy opts to pass integer error codes along the C-callable layer if some action fails in libindy. This allows wrapper consumers to gracefully detect errors that occurred in the SDK. A list of the errors can be found [here](../../../libindy/src/api/mod.rs). The comments give a short description of why the error may have occurred, but they won't help with debugging much. Alternatively, an excellent community maintained document would be to expand on the reasoning and solutions to why common error codes occur in certain situations.
```

Comment:
This information is outdated for the moment. Libindy now provides get_current_error method that returns deep information about error that includes descriptive method based on full error chain and backtrace.

Reply:
I haven't been apart of this work and I figured that there's more than one place where there's incorrect information in here. My hope was to get something started and have one of the core maintainers to make changes to it to accurately represent the layers. Ultimately the goal of this doc was to outline some of the things I learned while working in the codebase that could help others when they want to contribute to libindy. Would you or someone else on the team be able to make changes to this?